File: ProjectCS/GroupTable/StyledTextTableRenderer.cs
using System;
using System.Text;

namespace SessionResultsCS
{
    public class StyledTextTableRenderer : ITextTableRenderer
    {
        public const int MIN_COLUMN_WIDTH = 1;
        public const int RESULT_COLUMN_WIDTH = 3;
        public const int CELL_PADDING = 1;

        public static StyledTextTableRenderer Default { get; } = new StyledTextTableRenderer('+', '-', '|');


        // Символы, используемые для создания таблицы
        public char CrossSymbol { get; }
        public char HorizontalSymbol { get; }
        public char VerticalSymbol { get; }


        public StyledTextTableRenderer(char crossSymbol, char horizontalSymbol, char verticalSymbol)
        {
            CrossSymbol = crossSymbol;
            HorizontalSymbol = horizontalSymbol;
            VerticalSymbol = verticalSymbol;
        }


        public virtual string RenderTable(GroupTableData data)
        {
            int[] widths = CalculateColumnWidths(data);
            string separator = TableSeparator(widths, CrossSymbol, HorizontalSymbol);

            StringBuilder builder = new StringBuilder();

            builder.AppendLine(separator);
            builder.AppendLine(TableRow(data.TableHeader(), widths, VerticalSymbol));
            builder.AppendLine(separator);
            foreach (string[] cells in data.TableBodyToStrings(true))
            {
                builder.AppendLine(TableRow(cells, widths, VerticalSymbol));
            }
            builder.AppendLine(separator);
            builder.AppendLine(TableRow(data.TableDisciplineAverages(), widths, VerticalSymbol));
            builder.Append(separator);

            return builder.ToString();
        }


        protected static int[] CalculateColumnWidths(GroupTableData data)
        {
            int[] widths = new int[data.Disciplines.Length + GroupTableData.EXTRA_COLUMNS_COUNT];
            for (int i = 0; i < widths.Length; i++)
            {
                widths[i] = MIN_COLUMN_WIDTH;
            }
            widths[0] = Math.Max(widths[0], GroupTableData.FIO_COLUMN_HEADER.Length);
            foreach (Student student in data.Students)
            {
                widths[0] = Math.Max(widths[0], student.GetLastNameWithInitials().Length);
            }
            for (int i = 1; i <= data.Disciplines.Length; i++)
            {
                widths[i] = Math.Max(widths[i], RESULT_COLUMN_WIDTH);
            }
            widths[widths.Length - 1] = Math.Max(widths[widths.Length - 1], GroupTableData.AVERAGE_COLUMN_HEADER.Length);

            return widths;
        }

        protected static string TableSeparator(int[] widths, char crossSymbol, char horizontalSymbol)
        {
            StringBuilder builder = new StringBuilder();
            builder.Append(crossSymbol);
            foreach (int width in widths)
            {
                builder.Append(horizontalSymbol, width + CELL_PADDING * 2);
                builder.Append(crossSymbol);
            }
            return builder.ToString();
        }

        protected static string TableRow(string[] cells, int[] widths, char verticalSymbol)
        {
            StringBuilder builder = new StringBuilder();
            builder.Append(verticalSymbol);
            for (int i = 0; i < widths.Length; i++)
            {
                builder.Append(' ', CELL_PADDING);
                builder.Append(cells[i].PadLeft(widths[i]));
                builder.Append(' ', CELL_PADDING);
                builder.Append(verticalSymbol);
            }
            return builder.ToString();
        }
    }
}

File: Project/GroupTable/StyledTextTableRenderer.cpp
#include "GroupTable/StyledTextTableRenderer.hpp"

#include <sstream>


StyledTextTableRenderer::StyledTextTableRenderer(char crossSymbol, char horizontalSymbol, char verticalSymbol) :
	m_crossSymbol(crossSymbol), m_horizontalSymbol(horizontalSymbol), m_verticalSymbol(verticalSymbol) {}

StyledTextTableRenderer::StyledTextTableRenderer(const StyledTextTableRenderer& other) :
	m_crossSymbol(other.m_crossSymbol), m_horizontalSymbol(other.m_horizontalSymbol), m_verticalSymbol(other.m_verticalSymbol) {}

char StyledTextTableRenderer::GetCrossSymbol() const {
	return m_crossSymbol;
}

char StyledTextTableRenderer::GetHorizontalSymbol() const {
	return m_horizontalSymbol;
}

char StyledTextTableRenderer::GetVerticalSymbol() const {
	return m_verticalSymbol;
}

std::string StyledTextTableRenderer::RenderTable(const GroupTableData& data) {
	std::vector<int> widths = CalculateColumnWidths(data);
	std::string separator = TableSeparator(widths, m_crossSymbol, m_horizontalSymbol);

	std::stringstream ss;

	ss << separator << std::endl;
	ss << TableRow(data.TableHeader(), widths, m_verticalSymbol) << std::endl;
	ss << separator << std::endl;
	for (const std::vector<std::string> &row : data.TableBodyToStrings(true)) {
		ss << TableRow(row, widths, m_verticalSymbol) << std::endl;
	}
	ss << separator << std::endl;
	ss << TableRow(data.TableDisciplineAverages(), widths, m_verticalSymbol) << std::endl;
	ss << separator;

	return ss.str();
}

std::unique_ptr<StyledTextTableRenderer> StyledTextTableRenderer::GetDefault() {
	return std::make_unique<StyledTextTableRenderer>('+', '-', '|');
}

std::vector<int> StyledTextTableRenderer::CalculateColumnWidths(GroupTableData data) {
	std::vector<int> widths(data.GetDisciplines().size() + GroupTableData::EXTRA_COLUMNS_COUNT, MIN_COLUMN_WIDTH);
	widths[0] = std::max(widths[0], static_cast<int>(GroupTableData::FIO_COLUMN_HEADER.length()));
	for (auto& student : data.GetStudents()) {
		widths[0] = std::max(widths[0], static_cast<int>(student->GetLastNameWithInitials().length()));
	}
	for (int i = 1; i <= data.GetDisciplines().size(); i++) {
		widths[i] = std::max(widths[i], RESULT_COLUMN_WIDTH);
	}
	widths.back() = std::max(widths.back(), static_cast<int>(GroupTableData::AVERAGE_COLUMN_HEADER.length()));

	return widths;
}

std::string StyledTextTableRenderer::TableSeparator(const std::vector<int>& widths, char crossSymbol, char horizontalSymbol) {
	std::string separator(1, crossSymbol);
	for (int width : widths) {
		separator.append(width + CELL_PADDING * 2, horizontalSymbol);
		separator.push_back(crossSymbol);
	}
	return separator;
}

std::string StyledTextTableRenderer::TableRow(const std::vector<std::string>& cells, const std::vector<int>& widths, char verticalSymbol) {
	std::string row(1, verticalSymbol);
	for (int i = 0; i < widths.size(); i++) {
		row.append(CELL_PADDING, ' ');
		row.append(std::max(0, static_cast<int>(widths[i] - cells[i].length())), ' ');
		row.append(cells[i]);
		row.append(CELL_PADDING, ' ');
		row.push_back(verticalSymbol);
	}
	return row;
}
